vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,552 @@
+    pub log_version: u16,
+    pub version: u16,
+    pub log_length: u32,
+    pub log_offset: u64,
+}
+
+const HEADER_SIZE: u64 = 4096; // Each header size if 4 KiB
+const HEADER_SIGN: u32 = 0x6461_6568; //"head"
+const HEADER1_START: u64 = 64 * 1024; //header 1 start in Bytes
+const HEADER2_START: u64 = 128 * 1024; //header 2 start in Bytes
+
+impl Header {
+    /// Reads the Header structure from a reference VHDX file
+    pub fn new(f: &mut File, start: u64) -> Result<Header> {
+        // Read the whole header in to a buffer. We will need if for calculating checksum
+        let mut buffer = [0; HEADER_SIZE as usize];
+        f.seek(SeekFrom::Start(start)).map_err(Error::ReadHeader)?;
+        f.read_exact(&mut buffer).map_err(Error::ReadHeader)?;
+
+        let signature = LittleEndian::read_u32(&buffer[0..4]);
+        if signature != HEADER_SIGN {
+            return Err(Error::InvalidHeaderSign);
+        }
+
+        let checksum = LittleEndian::read_u32(&buffer[4..8]);
+        let new_checksum = match calculate_checksum(&mut buffer, size_of::<u32>()) {
+            Ok(r) => r,
+            Err(_) => return Err(Error::CalculateChecksum),
+        };
+
+        if checksum != new_checksum {
+            return Err(Error::InvalidChecksum(String::from("Header")));
+        }
+
+        let sequence_number = LittleEndian::read_u64(&buffer[8..16]);
+        let file_write_guid = LittleEndian::read_u128(&buffer[16..32]);
+        let data_write_guid = LittleEndian::read_u128(&buffer[32..48]);
+        let log_guid = LittleEndian::read_u128(&buffer[48..64]);
+        let log_version = LittleEndian::read_u16(&buffer[64..66]);
+        let version = LittleEndian::read_u16(&buffer[66..68]);
+        let log_length = LittleEndian::read_u32(&buffer[68..72]);
+        let log_offset = LittleEndian::read_u64(&buffer[72..80]);
+
+        Ok(Header {
+            signature,
+            checksum,
+            sequence_number,
+            file_write_guid,
+            data_write_guid,
+            log_guid,
+            log_version,
+            version,
+            log_length,
+            log_offset,
+        })
+    }
+
+    fn get_header_as_buffer(self, mut buffer: [u8; 80]) -> Header {
+        LittleEndian::write_u32(&mut buffer[0..4], self.signature);
+        LittleEndian::write_u32(&mut buffer[4..8], self.checksum);
+        LittleEndian::write_u64(&mut buffer[8..16], self.sequence_number);
+        LittleEndian::write_u128(&mut buffer[16..32], self.file_write_guid);
+        LittleEndian::write_u128(&mut buffer[32..48], self.data_write_guid);
+        LittleEndian::write_u128(&mut buffer[48..64], self.log_guid);
+        LittleEndian::write_u16(&mut buffer[64..66], self.log_version);
+        LittleEndian::write_u16(&mut buffer[66..68], self.version);
+        LittleEndian::write_u32(&mut buffer[68..72], self.log_length);
+        LittleEndian::write_u64(&mut buffer[72..80], self.log_offset);
+
+        self
+    }
```

Comment:
Ok so here you're trying to fill a buffer but you're not modifying `self`, so here is the correct syntax for doing such thing:
```suggestion
    fn get_header_as_buffer(&self, buffer: &mut [u8; 80]) {
        LittleEndian::write_u32(buffer[0..4], self.signature);
        LittleEndian::write_u32(buffer[4..8], self.checksum);
        LittleEndian::write_u64(buffer[8..16], self.sequence_number);
        LittleEndian::write_u128(buffer[16..32], self.file_write_guid);
        LittleEndian::write_u128(buffer[32..48], self.data_write_guid);
        LittleEndian::write_u128(buffer[48..64], self.log_guid);
        LittleEndian::write_u16(buffer[64..66], self.log_version);
        LittleEndian::write_u16(buffer[66..68], self.version);
        LittleEndian::write_u32(buffer[68..72], self.log_length);
        LittleEndian::write_u64(buffer[72..80], self.log_offset);
    }
```
You can't give ownership of the buffer to this function, otherwise the caller will never get the buffer back, which is the point of this function. That's why you need to pass a mutable reference.